radiography/src/main/java/radiography/Radiography.kt

```diff
@@ -0,0 +1,221 @@
+   * views of the currently focused window, if any.
+   */
+  @JvmStatic
+  fun scan(
+    rootView: View? = null,
+    includeTextViewText: Boolean = false,
+    textViewTextMaxLength: Int = Int.MAX_VALUE,
+    viewFilter: ViewFilter = ViewFilter.All
+  ): String = buildString {
+    if (includeTextViewText) {
+      check(textViewTextMaxLength >= 0) {
+        "textFieldMaxLength should be greater than 0, not $textViewTextMaxLength"
+      }
+    }
+
+    val rootViews = rootView?.let {
+      listOf(it)
+    } ?: WindowScanner.findAllRootViews()
+
+    val matchingRootViews = rootViews.filter(viewFilter::matches)
+
+    val config = Config(
+        includeTextViewText, textViewTextMaxLength, viewFilter
+    )
+
+    for (view in matchingRootViews) {
+      if (length > 0) {
+        append("\n")
+      }
+      val layoutParams = view.layoutParams
+      val title = (layoutParams as? WindowManager.LayoutParams)?.title?.toString()
+          ?: view.javaClass.name
+      appendln("$title:")
+
+      val startPosition = length
+      try {
+        appendln("window-focus:${view.hasWindowFocus()}")
+        config.scanRecursively(this, 0, 0, view)
+      } catch (e: Throwable) {
+        insert(
+            startPosition,
+            "Exception when going through view hierarchy: ${e.message}\n"
+        )
+      }
+    }
+  }
+
+  private class Config(
+    val includeTextViewText: Boolean,
+    val textViewTextMaxLength: Int,
+    val viewFilter: ViewFilter
+  )
+
+  private fun Config.scanRecursively(
+    result: StringBuilder,
+    depth: Int,
+    lastChildMask: Long,
+    view: View
+  ) {
+    @Suppress("NAME_SHADOWING")
+    var lastChildMask = lastChildMask
+    appendLinePrefix(result, depth, lastChildMask)
+    viewToString(result, view)
+    result.appendln()
+
+    if (view is ViewGroup) {
+      val lastNonSkippedChildIndex = findLastNonSkippedChildIndex(view)
+      val lastChildIndex = view.childCount - 1
+      for (index in 0..lastChildIndex) {
+        if (index == lastNonSkippedChildIndex) {
+          lastChildMask = lastChildMask or (1 shl depth).toLong()
+        }
+        val child = view.getChildAt(index)
+        // Never null on the main thread, but if called from another thread all bets are off.
+        if (child != null && viewFilter.matches(child)) {
+          scanRecursively(result, depth + 1, lastChildMask, child)
+        }
+      }
+    }
+  }
+
+  private fun Config.findLastNonSkippedChildIndex(viewGroup: ViewGroup): Int {
+    val lastChildIndex = viewGroup.childCount - 1
+    for (index in lastChildIndex downTo 0) {
+      val child = viewGroup.getChildAt(index)
+      if (viewFilter.matches(child)) {
+        return index
+      }
+    }
+    return -1
+  }
+
+  @TargetApi(CUPCAKE)
+  private fun Config.viewToString(
+    result: StringBuilder,
+    view: View
+  ) {
+    result.append("${view.javaClass.simpleName} { ")
+    if (view.id != -1 && view.resources != null) {
```

Comment:
Shouldn't this be `View.NO_ID` instead of `-1`? If not, it would still be better to extract our own private constant.